### PR TITLE
move numberOfAttempts to state from options

### PR DESCRIPTION
**Pull request by johnataylor**
- moves the numberOfAttempts property into the state from the options
- adds a typed/named property to the top level for parity with the typing/discovery C# implementation

### CODE REVIEW

**Comment by johnataylor** [comment]
also refer to https://github.com/Microsoft/botbuilder-dotnet/pull/1774

**Comment by benbrown** [comment]
is defaulting it to 0 the right thing to do?

**Comment by johnataylor** [comment]
Its interesting you should comment on that - the original was 0 based - the reason being one of the natural pieces of code to write was to look up the prompt test in an array and arrays are 0 based.
However, @gabog  changed this to 1 based. I gave feedback commenting on that change.
  
Anyhow I'm not that strongly of an opinion on this, but i think the 0 based array is something to consider. I mean this actually _is_ an index. (Unlike JavaScript month which is 0 based but isn't 😱 )

File: libraries/botbuilder-dialogs/tests/numberPrompt.test.js [hunk 1]
```diff
@@ -231,21 +231,21 @@ describe('NumberPrompt', function () {
         await adapter.send('Hello')
             .assertReply('Send me a zero')
```
Comment (gabog) [inline libraries/botbuilder-dialogs/tests/numberPrompt.test.js]: shouldn't the text be "send me a one"?

Comment (gabog) [reply libraries/botbuilder-dialogs/tests/numberPrompt.test.js]: done

**Comment by cleemullins** [comment]
The tests are failing. Build error is:

WARNING: QnAMaker test suite QNAKNOWLEDGEBASEID environment variable is not defined
WARNING: QnAMaker test suite QNAENDPOINTKEY environment variable is not defined
(node:1019) UnhandledPromiseRejectionWarning: AssertionError [ERR_ASSERTION]:  text === "Send me a zero"
    at defaultInspector (/Users/vsts/agent/2.150.0/work/1/s/libraries/botbuilder-core/lib/testAdapter.js:206:469)
    at waitForActivity (/Users/vsts/agent/2.150.0/work/1/s/libraries/botbuilder-core/lib/testAdapter.js:209:119)
    at Promise (/Users/vsts/agent/2.150.0/work/1/s/libraries/botbuilder-core/lib/testAdapter.js:209:339)
    at new Promise (<anonymous>)
    at TestFlow.previous.then (/Users/vsts/agent/2.150.0/work/1/s/libraries/botbuilder-core/lib/testAdapter.js:207:8)
(node:1019) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 49)
(node:1019) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
BotFrameworkAdapter.processActivity(): 400 ERROR - Error: BotFrameworkAdapter.parseRequest(): invalid request body.
BotFrameworkAdapter.processActivity(): 400 ERROR - Error: BotFrameworkAdapter.parseRequest(): missing activity type.
BotFrameworkAdapter.processActivity(): 401 ERROR - Error: failed auth
BotFrameworkAdapter.processActivity(): 500 ERROR - Error: bot exception
(node:1019) UnhandledPromiseRejectionWarning: Error: Unauthorized
    at MicrosoftAppCredentials.getToken (/Users/vsts/agent/2.150.0/work/1/s/libraries/botframework-connector/lib/auth/microsoftAppCredentials.js:35:643)
    at process._tickCallback (internal/process/next_tick.js:68:7)

**Pull request merged**